python/pyarrow/interchange/from_dataframe.py

```diff
@@ -0,0 +1,590 @@
+            if not null_count:
+                null_count = -1
+            array = pa.LargeStringArray.from_buffers(
+                length,
+                offset_pa_buffer,
+                data_pa_buffer,
+                validity_pa_buff,
+                null_count=null_count,
+                offset=offset,
+            )
+        else:
+            array = pa.Array.from_buffers(
+                pa.string(),
+                length,
+                [validity_pa_buff, offset_pa_buffer, data_pa_buffer],
+                offset=offset,
+            )
+    else:
+        array = pa.Array.from_buffers(
+            data_dtype,
+            length,
+            [validity_pa_buff, data_pa_buffer],
+            offset=offset,
+        )
+
+    # In case a bytemask was constructed with validity_buffer() call
+    # then we have to add the mask to the array
+    if bytemask is not None:
+        if not allow_copy:
+            raise RuntimeError(
+                "PyArrow Array will always be created out of a NumPy ndarray "
+                "and a copy is required which is forbidden by allow_copy=False"
+            )
+        return pa.array(array.to_pylist(), mask=bytemask)
+    else:
+        return array
+
+
+def validity_buffer(
+    validity_buff: BufferObject,
+    validity_dtype: Dtype,
+    describe_null: ColumnNullType,
+    offset: int = 0,
+) -> tuple[pa.Buffer, np.ndarray]:
+    """
+    Build a PyArrow buffer or a NumPy array from the passed buffer.
+
+    Parameters
+    ----------
+    validity_buff : BufferObject
+        Tuple of underlying validity buffer and associated dtype.
+    validity_dtype : Dtype
+        Dtype description as a tuple ``(kind, bit-width, format string,
+        endianness)``.
+    describe_null : ColumnNullType
+        Null representation the column dtype uses,
+        as a tuple ``(kind, value)``
+    offset : int, default: 0
+        Number of elements to offset from the start of the buffer.
+
+    Returns
+    -------
+    tuple
+        Tuple of a pyarrow buffer and a np.ndarray defining a mask.
+        Ony one of both can be not None.
+    """
+    null_kind, sentinel_val = describe_null
+    validity_kind, validity_bit_width, _, _ = validity_dtype
+    assert validity_kind == DtypeKind.BOOL
+
+    bytemask = None
+    if null_kind == ColumnNullType.USE_BYTEMASK:
+
+        bytemask = buffer_to_ndarray(validity_buff,
+                                     validity_bit_width,
+                                     offset)
+        if sentinel_val == 0:
+            bytemask = np.invert(bytemask)
+        validity_pa_buff = None
+
+    elif null_kind == ColumnNullType.USE_BITMASK and sentinel_val == 0:
+        validity_pa_buff = pa.foreign_buffer(
+            validity_buff.ptr, validity_buff.bufsize)
+    else:
+        raise NotImplementedError(
+            f"{describe_null} null representation is not yet supported.")
+
+    return validity_pa_buff, bytemask
+
+
+def buffer_to_ndarray(
```

Comment:
Why do we need this helper for a bytemask? I would assume we can take the same approach as for int8 arrays? (which would go through pa.foreign_buffer and pa.Array.from_buffers)